### PR TITLE
Volume segmentation and derecursification

### PR DESCRIPTION
**Pull request by sloriot**
## Summary of Changes

This PR adds the function `volume_connected_component()` that associates to each face a volume component id.

This function is used as the implementation of `does_bound_a_volume()` and `orient_to_bound_a_volume()`
## Release Management

* Affected package(s): PMP
* Feature/Small Feature (if any): [here](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Volume_connected_components) **Pre-approval date:** 2020/03/17
* Link to compiled documentation (obligatory for small feature):
  * [User man 1](https://cgal.geometryfactory.com/~sloriot/doc/VolSeg/r1/Polygon_mesh_processing/#title39
)
  * [User man 2](https://cgal.geometryfactory.com/~sloriot/doc/VolSeg/r1/Polygon_mesh_processing/#title54)
  * [Ref Man](https://cgal.geometryfactory.com/~sloriot/doc/VolSeg/r1/Polygon_mesh_processing/group__PMP__orientation__grp.html#ga45f497ccb2caee5d2222b6057901506f)
* License and copyright ownership: GF

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h [hunk 1]
```diff
@@ -468,23 +471,68 @@ enum Volume_error_code { VALID_VOLUME, ///< The set of faces bounds a volume
                          SINGLE_CONNECTED_COMPONENT ///< The set of faces consists of all the faces of the mesh and no further test was done.
                        };
 namespace internal {
-template<class RefToVector>
-void copy_error_codes(std::vector<Volume_error_code>& error_codes_in,
-                      RefToVector ref_wrapper)
+
+template<class T, class RefToContainer>
+void copy_container_content(std::vector<T>& vec, RefToContainer ref_wrapper)
```
Comment (lrineau) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h]: It seems `vec` could be `const`.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h [hunk 2]
```diff
@@ -542,6 +456,810 @@ void orient(TriangleMesh& tm)
...
+ * assigns to each face of `tm` an id corresponding to the volume connected component
+ * it contributes to define.
+ *
+ * Using the adjacency relation of two faces along an edge, a triangle mesh can be split
+ * into connected components (*surface components* in the following).
+ * A surface component without boundary separates the 3D space into an infinite and
+ * a finite volume. We say that the finite volume is <i>enclosed</i> by this surface
+ * component.
+ *
+ * The volume connected components (*volume components* in the following) are defines as follow:
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h]:
```suggestion
 * The volume connected components (*volume components* in the following) are defined as follows:
```

**Comment by sloriot** [comment]
@maxGimeno please refresh the branch. @MaelRL did a live review some time ago and seems the API could be submit in a SF as is.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h [hunk 3]
```diff
@@ -535,6 +449,819 @@ void orient(TriangleMesh& tm)
...
+        else
+          error_codes.push_back(VALID_VOLUME);
+      }
+
+      for(std::size_t ncc_id : nested_cc_per_cc[cc_id])
+      {
+        if ( nesting_levels[ncc_id]==nesting_levels[cc_id]+1 )
+        {
+          cc_handled.set(ncc_id);
+          if (!ignore_orientation_of_cc)
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h]: three indentation styles :>

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h [hunk 4]
```diff
@@ -535,6 +449,819 @@ void orient(TriangleMesh& tm)
...
+ * \see `CGAL::Polygon_mesh_processing::orient_to_bound_a_volume()`
+ */
+template <class TriangleMesh, class NamedParameters>
+bool does_bound_a_volume(const TriangleMesh& tm, const NamedParameters& np)
+{
+  typedef boost::graph_traits<TriangleMesh> GT;
+  typedef typename GT::face_descriptor face_descriptor;
+
+  if (!is_closed(tm)) return false;
+  if (!is_triangle_mesh(tm)) return false;
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h]: should these be preconditions instead?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h [hunk 5]
```diff
@@ -535,6 +449,819 @@ void orient(TriangleMesh& tm)
...
+ *   \cgalParamEnd
+ *   \cgalParamBegin{face_connected_component_map}
+ *     a property map filled by this function and that will contain for each face the id
+ *     of its surface component in the range `[0, number of surface components - 1[`.
+ *   \cgalParamEnd
+ *   \cgalParamBegin{volume_inclusions}
+ *     a `reference_wrapper` (either from `boost` or the standard library) containing
+ *     a reference to an object that must be a model of the `BackInsertionSequence` concept,
+ *     with a value type being a model of `BackInsertionSequence` of `std::size_t`,
+ *     both types having the functions `reserve()` and `push_back()`.
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h]: I'm adding a reserve() that will work for containers that do not have reserve, in https://github.com/CGAL/cgal/pull/4418/files#diff-f1e74bf64644af37d70f782ac4cb1ee9. Maybe use that instead?

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/NamedParameters.txt [hunk 6]
```diff
@@ -422,6 +422,108 @@ flips to recover good shapes, unless specified otherwise (see below).
...
+\b Default: None
+\cgalNPEnd
+
+\cgalParamBegin{connected_component_id_to_volume_id}
+a `reference_wrapper` (either from `boost` or the standard library) containing
+a reference to an object that must be a model of the `BackInsertionSequence` concept,
+with a value_type being `std::size_t`.
+The size of the container is exactly the number of connected components.
+For each connected component identified using its id `ccid`, the id of the volume it contributes
+to describe is the value at the position `ccid` in the parameter container.
```
[the remark this answers is not in the document]
Comment (sloriot) [reply Polygon_mesh_processing/doc/Polygon_mesh_processing/NamedParameters.txt]:
```suggestion
to is the value at the position `ccid` in the parameter container.
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/NamedParameters.txt [hunk 7]
```diff
@@ -422,6 +422,108 @@ flips to recover good shapes, unless specified otherwise (see below).
...
+\b Default: None
+\cgalParamEnd
+
+\cgalParamBegin{nesting_levels}
+a `reference_wrapper` (either from `boost` or the standard library) containing
+a reference to an object that must be a model of the `BackInsertionSequence` concept,
+with a value_type being `std::size_t`.
+The size of the container is exactly the number of connected components.
+For each connected component identified using its id `ccid`, the container contains the number of
+connected components containing on its bounded side this component.
```
[the remark this answers is not in the document]
Comment (sloriot) [reply Polygon_mesh_processing/doc/Polygon_mesh_processing/NamedParameters.txt]:
```suggestion
connected components containing on their bounded side this component.
```

[the remark this answers is not in the document]
Comment (sloriot) [reply Polygon_mesh_processing/doc/Polygon_mesh_processing/NamedParameters.txt]: should be clearer with the right possessive adjective.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h [hunk 8]
```diff
@@ -535,6 +449,819 @@ void orient(TriangleMesh& tm)
...
+ * - If all the faces of `tm` are part of the same surface component, then
+ *   no further test (open/closed, self-intersecting) is done and all faces
+ *   are assigned to the same volume components.
+ * - If `do_orientation_tests` is set to `true` and the surface components that are
+ *   outside all enclosed volumes are inward oriented, they are then considered as holes
+ *   of the unbounded volume (that has no outer boundary)
+ *
+ * A property map for each of `CGAL::face_index_t` and `CGAL::vertex_point_t`
+ * must be either available as an internal property map
+ * of `tm` or provided as one of the \ref pmp_namedparameters "Named Parameters".
```
[the remark this answers is not in the document]
Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h]: I don't want to merge a PR that is not yet integrated (a pain when using git-filter branch), let see how things evolve.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h (hunk 5 above)
Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h]: same remark as above

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-108.shtml

**Pull request merged**